solr/core/src/java/org/apache/solr/update/SolrCoreState.java

```diff
@@ -86,7 +103,45 @@ public boolean decrefSolrCoreState(IndexWriterCloser closer) {
     }
     return close;
   }
-  
+
+  /**
+   * Pauses all update requests to this core and waits (indefinitely) for all in-flight
+   * update requests to finish
+   */
+  public void pauseUpdatesAndAwaitInflightRequests() {
+    if (pauseUpdateRequests.compareAndSet(false, true)) {
+      inflightUpdatesCounter.register();
+      inflightUpdatesCounter.arriveAndAwaitAdvance();
```

Comment:
This is in the shutdown path, should we add a timeout? What happens if a disk is slow on write and something else times out underneath us?

Reply:
Good point. I have added a small timeout (2500 ms) with the same value as the one used inside RecoveryStrategy for leader to finish in-flight requests. I think this default should be enough but it can be changed using a sysprop if needed.